### PR TITLE
Add "Related work" section to Title module

### PR DESCRIPTION
**Pull request by NitinBhasneria**
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

Fixes: #470 

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `grunt jasmine`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] PR body includes `fixes #0000`-style reference to original issue #
* [x] ask `@publiclab/reviewers` for help, in a comment below

Before:
title2

After:
title1

In this PR I have added a function for clicking in the `input` which will also show the suggested posts. I have replaced the `focusout` function with `mouseleave`.

### CODE REVIEW

**Comment by NitinBhasneria** [comment]
cc: @VladimirMikulic @shreyaa-sharmaa @Shulammite-Aso @jywarren @ebarry @keshav234156

**Comment by NitinBhasneria** [comment]
titleAdded
In this, I have added the `Added Work` Div.

**Comment by jywarren** [comment]
Hi, this looks great, could we remove the "alert" language as discussed in #470 and also change `added works` to `Related posts`?

Also just noticing perhaps the button classes have to be updated to display as buttons in this view? Thank you!

**Comment by NitinBhasneria** [comment]
> Hi, this looks great, could we remove the "alert" language as discussed in #470 and also change `added works` to `Related posts`?
> 
> Also just noticing perhaps the button classes have to be updated to display as buttons in this view? Thank you!

Will this surely, currently the PR is in progress

**Comment by NitinBhasneria** [comment]
> Hi, this looks great, could we remove the "alert" language as discussed in #470 and also change `added works` to `Related posts`?
> 
> Also just noticing perhaps the button classes have to be updated to display as buttons in this view? Thank you!

@jywarren just confused about which alert language you are talking about. Sorry for that.

**Comment by shreyaa-sharmaa** [comment]
Just to clarify, will the related posts be hidden on `mouseleave` or `focusout` or visible all the time?

**Comment by NitinBhasneria** [comment]
> Just to clarify, will the related posts be hidden on `mouseleave` or `focusout` or visible all the time?

Will be hidden on mouseleave.

**Comment by NitinBhasneria** [comment]
@jywarren @cesswairimu @VladimirMikulic @shreyaa-sharmaa @keshav234156 @Shulammite-Aso @ebarry 
I think all the required tasks are completed. Please have a look and review this.

titlecomplete

**Comment by cesswairimu** [comment]
wow this looks great :tada: is the same available when editing?

**Comment by NitinBhasneria** [comment]
> wow this looks great  is the same available when editing?

Yes, we can edit anytime.

**Comment by cesswairimu** [comment]
Awesome :100:  I also see bunch of changes on this file `dist/PublicLab.Editor.js` kindly confirm you need those or discard the changes. Thanks

**Comment by NitinBhasneria** [comment]
> Awesome  I also see bunch of changes on this file `dist/PublicLab.Editor.js` kindly confirm you need those or discard the changes. Thanks

No, These changes are only due to grunt build. The changes will work without it. Also, it won't affect anything as grunt will build the dist every time.
Thanks.

**Comment by cesswairimu** [comment]
aha, gotcha :ok: , if its not much trouble you could remove the changes you made on the file just for accuracy...its showing +9k changes. Thanks

**Comment by NitinBhasneria** [comment]
Done 👍. Thanks

File: src/modules/PublicLab.TitleModule.Related.js [hunk 1]
```diff
@@ -35,28 +38,73 @@ module.exports = function relatedNodes(module) {
...
+    });
 
     relatedResults.slice(0, 8).forEach(function(result) {
-
-      relatedEl.append('<div class="result result-' + result.id + '" style="margin: 3px;"><a class="btn btn-xs btn-default add-tag"><i class="fa fa-plus-circle"></i> Add</a> <a class="title"></a> by <a class="author"></a></div>');
-      relatedEl.find('.result-' + result.id + ' .title').html(result.title);
-      relatedEl.find('.result-' + result.id + ' .title').attr('href', result.url);
-      relatedEl.find('.result-' + result.id + ' .author').html('@' + result.author);
-      relatedEl.find('.result-' + result.id + ' .author').attr('href', '/profile/' + result.author);
+      var showRealted = false;
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:86]: `showRelated`

Comment (NitinBhasneria) [reply src/modules/PublicLab.TitleModule.Related.js:86]: @VladimirMikulic  I have applied the preetier format on the changes nothing has changed except some space after `function ()` but we are using `function()` in all the places.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 2]
```diff
@@ -35,28 +38,73 @@ module.exports = function relatedNodes(module) {
...
-    relatedEl.append('<p class="ple-help">Does your work relate to one of these? Click to alert those contributors.</p><hr style="margin: 4px 0;" />');
+    relatedEl.append('<p class="ple-help">Does your work relate to one of these? Click to tag the related work.</p><hr style="margin: 4px 0;" />');
+    module.el.find('.ple-module-content').append('<div style="display:none;" class="ple-title-added"></div>');
+    addedRelatedEl = module.el.find('.ple-title-added');
 
   }
 
+  function showAdded(addedResult, relatedResults) {
+    addedRelatedEl.append('<div class="addedresult addedresult-' + addedResult.id + '" style="margin: 3px;"><a class="btn btn-xs btn-default remove-tag"><i class="fa fa-times-circle"></i> Remove</a> <a class="addedtitle"></a> by <a class="addedauthor"></a></div>');
+    addedRelatedEl.find('.addedresult-' + addedResult.id + ' .addedtitle').html(addedResult.title);
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:49]: `'.addedresult-' + addedResult.id` is repeated a lot. It should be in a variable.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 3]
```diff
@@ -35,28 +38,73 @@ module.exports = function relatedNodes(module) {
...
+    addedRelatedEl.find('.addedresult-' + addedResult.id + ' .addedtitle').attr('href', addedResult.url);
+    addedRelatedEl.find('.addedresult-' + addedResult.id + ' .addedauthor').html('@' + addedResult.author);
+    addedRelatedEl.find('.addedresult-' + addedResult.id + ' .addedauthor').attr('href', '/profile/' + addedResult.author);
+
+    $('.addedresult-' + addedResult.id + ' .remove-tag').click(function() {
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:54]: This block can be refactored in a standalone function and documented since it's quite large.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 4]
```diff
@@ -35,28 +38,73 @@ module.exports = function relatedNodes(module) {
...
+      show(relatedResults);
+      if(addedRelatedPost==='addedTitles') {
+        $('.related-post').remove();
+      }
+    });
+  }
+
   // expects array of results in format:
   // { id: 3, title: 'A third related post', url: '/', author: 'bsugar'}
   function show(relatedResults) {
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:70]: `showRelatedResults` would be a better function name here.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 5]
```diff
@@ -35,28 +38,73 @@ module.exports = function relatedNodes(module) {
...
-      relatedEl.find('.result-' + result.id + ' .title').attr('href', result.url);
-      relatedEl.find('.result-' + result.id + ' .author').html('@' + result.author);
-      relatedEl.find('.result-' + result.id + ' .author').attr('href', '/profile/' + result.author);
+      var showRelated = false;
+      if(!(addedRelatedPost.includes(result.id))) {
+        showRelated=true;
+      }
+      if(showRelated) {
+        relatedEl.append('<div class="result result-' + result.id + '" style="margin: 3px;"><a class="btn btn-xs btn-default add-tag"><i class="fa fa-plus-circle"></i> Add</a> <a class="title"></a> by <a class="author"></a></div>');
+        relatedEl.find('.result-' + result.id + ' .title').html(result.title);
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:92]: `'.result-' + result.id` can be refactored in a variable.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 6]
```diff
@@ -81,15 +129,28 @@ module.exports = function relatedNodes(module) {
...
+        relatedResultShow = true;
         fetchRelated(show);
       }
  
     });
+
+    $(module.el).find('input').click(function(e) {
  
-    $(module.el).find('input').focusout(function(e) {
+      if (module.options.suggestRelated && relatedResultShow) {
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:141]: `module.options.suggestRelated && relatedResultShow` it's better to refactore expressions in variables. It's more readable.

**Comment by NitinBhasneria** [comment]
@VladimirMikulic have a look.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 7]
```diff
@@ -81,15 +132,29 @@ module.exports = function relatedNodes(module) {
...
+        addedRelatedEl.fadeIn();
+        relatedResultShow = true;
+        fetchRelated(showRelatedResult);
       }
  
     });
+
+    $(module.el).find('input').click(function(e) {
+      var showResults = false;
+      showResults = module.options.suggestRelated && relatedResultShow;
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:144]:
```javascript
   var showResults = false;
   showResults = module.options.suggestRelated && relatedResultShow;
```

Why not declaration and initialization on the same line? It doesn't make sense to initialize this to `false` and then override it?

File: src/modules/PublicLab.TitleModule.Related.js [hunk 8]
```diff
@@ -81,15 +132,29 @@ module.exports = function relatedNodes(module) {
  
       if (module.options.suggestRelated) {
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:133]: This `if` block is present on the line 145 as well. They include 75% of the code in common. Can be refactored in a standalone function.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 9]
```diff
@@ -35,39 +38,87 @@ module.exports = function relatedNodes(module) {
...
     relatedResults.slice(0, 8).forEach(function(result) {
-
-      relatedEl.append('<div class="result result-' + result.id + '" style="margin: 3px;"><a class="btn btn-xs btn-default add-tag"><i class="fa fa-plus-circle"></i> Add</a> <a class="title"></a> by <a class="author"></a></div>');
-      relatedEl.find('.result-' + result.id + ' .title').html(result.title);
-      relatedEl.find('.result-' + result.id + ' .title').attr('href', result.url);
-      relatedEl.find('.result-' + result.id + ' .author').html('@' + result.author);
-      relatedEl.find('.result-' + result.id + ' .author').attr('href', '/profile/' + result.author);
+      var showRelated = false;
+      if (!(addedRelatedPost.includes(result.id))) {
+        showRelated=true;
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:93]: Format the file with Prettier.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 10]
```diff
@@ -35,39 +38,87 @@ module.exports = function relatedNodes(module) {
...
+  function showRelatedResult(relatedResults) { 
     relatedEl.find('.result').remove();
+    // If tag removed from TagModule
+    editor.tagsModule.el.find('input').on('tokenfield:removedtoken', function(e) {
+      var id = (e.attrs.value).replace('response:', '');
+      if (addedRelatedPost.includes(id)) {
+        addedRelatedPost = addedRelatedPost.replace(id, '');
+        $('.addedresult-' + id).remove();
+        showRelatedResult(relatedResults);
+        if (addedRelatedPost==='addedTitles') {
```
Comment (VladimirMikulic) [inline src/modules/PublicLab.TitleModule.Related.js:84]: Formatting needed.

**Comment by VladimirMikulic** [comment]
Hi @govindgoel. The commits are squashed upon merging PR, so it is not mandatory for @NitinBhasneria to do it.

File: src/modules/PublicLab.TitleModule.Related.js (hunk 7 above)
Comment (NitinBhasneria) [reply src/modules/PublicLab.TitleModule.Related.js]: Done

**Comment by VladimirMikulic** [comment]
@NitinBhasneria ignoring changes won't get your PR merged. The file is still not formatted correctly. Do I have to point everything manually? Use a tool. Code is still not documented. Follow the best practises. I can't approve something that will be a huge maintenance problem in the future.

**Comment by NitinBhasneria** [comment]
> @NitinBhasneria ignoring changes won't get your PR merged. The file is still not formatted correctly. Do I have to point everything manually? Use a tool. Code is still not documented. Follow the best practises. I can't approve something that will be a huge maintenance problem in the future.

Yup, I have formatted the file with prettier extension. If you find any line not looking good after the prettier than please let me know. For documentation I am adding some comments. Thank You.

**Comment by NitinBhasneria** [comment]
@jywarren have a look and give your review.😊

**Comment by jywarren** [comment]
Hi all. @VladimirMikulic, please remember to be friendly at all times -- it's a very important way that we show respect to one another. I know it can be frustrating sometimes, and myself I sometimes have to take a deep breath and step away for a moment. We know you mean well, and appreciate your help!

@NitinBhasneria you're doing great. please do try to go through each change so that reviewing goes smoothly. 

Thank you both for your patience in this PR, which is ending up a bit more complex than we'd thought! Appreciation for you both and @govindgoel and @cesswairimu for your support. 

❤️

File: src/modules/PublicLab.TitleModule.Related.js [hunk 11]
```diff
@@ -26,72 +26,141 @@ Results should be in following JSON format:
...
+  // showResult() will append the required tags in the particular container
+  function showResults(resultContainer, result, resultClass) {
+    resultContainer.find(resultClass + ' .title').html(result.title);
+    resultContainer.find(resultClass + ' .title').attr('href', result.url);
+    resultContainer.find(resultClass + ' .author').html('@' + result.author);
+    resultContainer.find(resultClass + ' .author').attr('href', '/profile/' + result.author);
+  }
+
+  function showAdded(addedResult, relatedResults) {
+    addedRelatedEl.append('<div class="addedresult addedresult-' + addedResult.id + '" style="margin: 3px;"><a class="btn btn-xs btn-default remove-tag"><i class="fa fa-times-circle"></i> Remove</a> <a class="title"></a> by <a class="author"></a></div>');
```
Comment (jywarren) [inline src/modules/PublicLab.TitleModule.Related.js:56]: Are the button outlines not appearing (despite `btn btn-xs btn-default) because the example files are on Bootstrap 3 and the `plots2` project is on Bootstrap 4? Is there a way to use CSS to achieve decent styling on both (like including `btn-link` or something) or should we just solve this in a new PR which bumps the Editor's Bootstrap to v4? I could have this wrong but just trying to troubleshoot.

https://getbootstrap.com/docs/4.4/components/buttons/

https://getbootstrap.com/docs/3.3/css/#buttons

Thank you!

**Comment by VladimirMikulic** [comment]
Hi @jywarren. I apologise if I had done something wrong. Others may not like my approach, but I tend to have almost everything "pixel-perfect". I understand if you would want me to step aside. No hard feelings. :+1:

File: src/modules/PublicLab.TitleModule.Related.js (hunk 11 above)
Comment (NitinBhasneria) [reply src/modules/PublicLab.TitleModule.Related.js:56]: I will do it with using the css.

**Comment by NitinBhasneria** [comment]
> Hi @jywarren. I apologise if I had done something wrong. Others may not like my approach, but I tend to have almost everything "pixel-perfect". I understand if you would want me to step aside. No hard feelings.

There is no problem from my side. I like your reviews. No hard feeling @VladimirMikulic. Also, I truly understand your work, Thanks you.

**Comment by SidharthBansal** [comment]
@VladimirMikulic I think you are helpful but please use a helping friendly and understanding tone. Like 
>> ignoring changes won't get your PR merged. 
>> Do I have to point everything manually?

Reviewers/mentors are for guiding the contributors. You are contributing since much longer time than nitin. That's great. So, you know much more than him. So, there will be many times you should help the contributors.

>> The file is still not formatted correctly.  Use a tool. Code is still not documented. Follow the best practises. I can't approve something that will be a huge maintenance problem in the future.

Can be changed to
 - Please format the file by using a tool named .... . Kindly add documentation following the best practices. Documentation and formatting is needed for better maintainability of the code so that the future developer can easily understand the codebase and commit changes. 

Please use `constructive criticism in a requesting tone/friendly tone`. 


>> I understand if you would want me to step aside

We encourage you to continue with public lab. You did quite well till now and we hope you will do better.

In case you have any concerns, feel free to ping me on gitter personal chat @VladimirMikulic @NitinBhasneria . I remain busy with my job so not able to give time here. Apologies.

**Comment by jywarren** [comment]
@VladimirMikulic i echo Sidharth's excellent response. We are very happy to have you in the community and your detailed support for folks is noted and VERY appreciated. The kinds of in-depth reviews you are providing are excellent and no complaints there. 

I think Sidharth has really captured how we hope to have kind and friendly tone in our interactions, not only because it's good to be kind, but also because it helps us all feel good about our work and collaborations and so, work together better. It does sometimes take a little longer to add a friendly quality to our comments. But it has a wonderful immediate and long-term effect on our cohesion and mutual support as a community and so is worth it every time 🙌  I also understand if you need to pace yourself on reviewing in order to ensure you can bring your best self to each session. This is natural -- take care of yourself, too! We appreciate every minute. 

I'm also always available to chat via Gitter if you'd like to discuss more. Keep up the great 'pixel-perfect' work! ❤️

**Comment by VladimirMikulic** [comment]
Thank you all!

**Comment by NitinBhasneria** [comment]
@jywarren have a look 
btn

**Comment by NitinBhasneria** [comment]
> Now that we have Jest testing, would you be able to add a couple simple tests to preserve and protect this behavior? Thank you!

OK, will do it for sure.
Also, wanted to ask is there any problem if I add the test in `/specs` and run by `grunt jasmine`

**Comment by jywarren** [comment]
I think the easiest may be to write Jest tests, but if you can do it in Jasmine and it works for you, sounds great! Jasmine is a "simulated" browser environment whereas Jest really runs a headless browser.

**Comment by NitinBhasneria** [comment]
@jywarren testa are added.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 12]
```diff
@@ -25,59 +25,142 @@ Results should be in following JSON format:
  */
 module.exports = function relatedNodes(module) {
   var relatedEl;
+  var addedRelatedEl;
+  var addedRelatedPost = 'addedTitles';
```
Comment (sagarpreet-chadha) [inline src/modules/PublicLab.TitleModule.Related.js:29]: Why string? Use an array. 
If we use string, it will cause bug when id is two digit or more, right? 
"123" matches 12 and 23

File: src/modules/PublicLab.TitleModule.Related.js [hunk 13]
```diff
@@ -25,59 +25,142 @@ Results should be in following JSON format:
...
+      // For adding tags.
 
       $('.result-' + result.id + ' .add-tag').click(function() {
         editor.tagsModule.el.find('input').tokenfield('createToken', 'response:' + result.id);
+        if (addedRelatedPost === 'addedTitles') {
+          addedRelatedEl.append('<hr class="related-post" style="margin: 4px 0;" /><p class="ple-help related-post">Related Posts</p>');
+        }
         // pending https://github.com/publiclab/plots2/issues/646
         // editor.tagsModule.el.find('input').tokenfield('createToken', 'notify:' + result.author);
+        addedRelatedPost += result.id;
```
Comment (sagarpreet-chadha) [inline src/modules/PublicLab.TitleModule.Related.js:111]: append to an array not string here.

File: src/modules/PublicLab.TitleModule.Related.js [hunk 14]
```diff
@@ -25,59 +25,142 @@ Results should be in following JSON format:
...
+      }
+    });
 
     relatedResults.slice(0, 8).forEach(function(result) {
-      relatedEl.append('<div class="result result-' + result.id + '" style="margin: 3px;"><a class="btn btn-xs btn-default add-tag"><i class="fa fa-plus-circle"></i> Add</a> <a class="title"></a> by <a class="author"></a></div>');
-      relatedEl.find('.result-' + result.id + ' .title').html(result.title);
-      relatedEl.find('.result-' + result.id + ' .title').attr('href', result.url);
-      relatedEl.find('.result-' + result.id + ' .author').html('@' + result.author);
-      relatedEl.find('.result-' + result.id + ' .author').attr('href', '/profile/' + result.author);
+      var showRelated = false;
```
Comment (sagarpreet-chadha) [inline src/modules/PublicLab.TitleModule.Related.js:92]: No need for this variable, directly use if condition of line 93 without negation

File: src/modules/PublicLab.TitleModule.Related.js [hunk 15]
```diff
@@ -25,59 +25,142 @@ Results should be in following JSON format:
...
   // expects array of results in format:
   // { id: 3, title: 'A third related post', url: '/', author: 'bsugar'}
-  function show(relatedResults) {
+  function showRelatedResult(relatedResults) {
     relatedEl.find('.result').remove();
+    // If tag removed from TagModule
+    editor.tagsModule.el.find('input').on('tokenfield:removedtoken', function(e) {
+      var id = (e.attrs.value).replace('response:', '');
+      if (addedRelatedPost.includes(id)) {
+        addedRelatedPost = addedRelatedPost.replace(id, '');
```
Comment (sagarpreet-chadha) [inline src/modules/PublicLab.TitleModule.Related.js:82]: use array and splice

**Comment by NitinBhasneria** [comment]
@sagarpreet-chadha have a look!

**Comment by jywarren** [comment]
Looking great! I'll try it out in GitPod too!

File: spec/javascripts/title_module_spec.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+  });
+
+  it("value, valid, key", function() {
+    expect(module).not.toBeUndefined();
+    expect(module.key).toBe('title');
+
+    expect(module.value()).not.toBe(false);
+    expect(module.value()).toBe('');
+
+    module.value('My new post:');
```
Comment (jywarren) [inline spec/javascripts/title_module_spec.js:21]: Here, I'm wondering if we should test specifically the idea of adding/removing items, and confirming that they appear/disappear in the right places? See how @Shulammite-Aso has used Jest to create a set of conditions and then is testing whether the outcome is correct, here? 

https://github.com/jywarren/woofmark/pull/15/files#diff-635478b67938477db9d8a9bc88585289R10

Thank you!

**Comment by jywarren** [comment]
This looks ok, but I'm not getting any results in GitPod... @NitinBhasneria am i supposed to be able to type in "test" there?

image


```
Change detected /workspace/PublicLab.Editor/dist/PublicLab.Editor.js
GET /examples/undefined 404 0.797 ms - 157
GET /examples/undefined 404 1.134 ms - 157
```

I'm wondering if we need to set up a mocked response in the demo, like a JSON file we can point the demo at so we can show it responding? Then the request could go to, for example, `/examples/relatedTitles.json`?

File: spec/javascripts/title_module_spec.js (hunk 16 above)
Comment (NitinBhasneria) [reply spec/javascripts/title_module_spec.js:21]: Will do it soon thanks.

**Comment by NitinBhasneria** [comment]
@jywarren I have tried writing the test for `related title module` but as related titles come from the already published post, in the test it is not rendering anything. I tried publishing the dummy post in the test but got nothing.
@Shulammite-Aso @shreyaa-s @keshav234156 any idea how we can write the test for this?

**Comment by sagarpreet-chadha** [comment]
Hey Nitin, you have to mock related titles. So if you will not pass fetchRelated in options it will take the hardcoded array of related titles, right? So mocking in this case is already happening.
So now you need to.write test exactly as what you are doing in the GIF above, you can also mock mouse position using jest. Thanks!

**Comment by NitinBhasneria** [comment]
> Hey Nitin, you have to mock related titles. So if you will not pass fetchRelated in options it will take the hardcoded array of related titles, right? So mocking in this case is already happening.
> So now you need to.write test exactly as what you are doing in the GIF above, you can also mock mouse position using jest. Thanks!

Yea, but the fetchRelated will be empty as there are no post in the existing database in our system.

**Comment by sagarpreet-chadha** [comment]
Nitin, there is 2 things here, one is UI part (which is PL.editor) and another is the server which handles the  api's generated from UI.
We are interested here only in the UI part, there is no server in picture yet. So we need to mock these api calls made from UI by returning a hardcoded array of related titles which we will pass as options into PL.editor intantiated object.
Now how to mock api is explained in the jest documentation, let me know if this not clear. Thanks 😄

**Comment by NitinBhasneria** [comment]
@sagarpreet-chadha I am confused about mocking the function as `showRelatedResult` is already a function inside `relatedNodes()` in https://github.com/publiclab/PublicLab.Editor/blob/39bd6e979920a5cb07dc318450fa19dcb68e36d7/src/modules/PublicLab.TitleModule.Related.js#L26. So it's like mocking a function inside a function. How should we do this?

**Comment by NitinBhasneria** [comment]
@sagarpreet-chadha as far I have looked I got to know we cannot export the function inside the function if it is in same module https://stackoverflow.com/questions/50854440/spying-on-an-imported-function-that-calls-another-function-in-jest/50855968#50855968. And so, I can't find a way to do this. @shreyaa-s @Shulammite-Aso @keshav234156 any suggestion?

**Comment by NitinBhasneria** [comment]
OK, I have got the idea from @jywarren. I will make the `HTML` file in examples folder for tags and title module(named 'examples/titleReladted.js'. This HTML file will be attached to `titleRelated.js`(again a new file for title module) this file will contain the js content of adding the related module. And then we can test the `examples/titleRelated.html`.
@jywarren @sagarpreet-chadha let me know this will be good or not?

**Comment by sagarpreet-chadha** [comment]
Yes sure, a new `titleRelated.js` looks fine (btw this is what mocking means)

**Comment by NitinBhasneria** [comment]
> Yes sure, a new `titleRelated.js` looks fine (btw this is what mocking means)

Ok then, I will do the same.

**Comment by NitinBhasneria** [comment]
> Yes sure, a new `titleRelated.js` looks fine (btw this is what mocking means)

@sagarpreet-chadha with this approach too, we can't show everything working. I mean this is how it will work: we will manually add the mock `add related tags`. And on clicking these tags the `remove-tag` will be generated. and on clicking the `remove-tag` the `remove-tag` will be removed and the `add-tag` will be added again. Does this look good?


Also, as tests are to test the real add related module that is: src/modules/PublicLab.TitleModule.Related.js but by this method we are adding manually and not testing the file. Its really complicated now 😅.
@jywarren please give your views.

**Comment by NitinBhasneria** [comment]
Update:
Tried doing the same with the jest test, but still, we cant import the function inside the function. Up to now the only way to do this is to copy the code in test and then check it as I have mentioned in the above comment.

cc: @jywarren  @sagarpreet-chadha @cesswairimu

**Comment by jywarren** [comment]
Hi! great working happening here. Just noting this; 

> we will manually add the mock add related tags. And on clicking these tags the remove-tag will be generated. and on clicking the remove-tag the remove-tag will be removed and the add-tag will be added again. Does this look good?

This sounds perfect. In testing, really we can do tests of each discrete component, which confirm that they react to the rest of the system in the correct way. in this way we can "mock" a scenario or set of expected responses for a subcomponent of the whole, and the tests confirm that this subcomponent reacts as expected. 

The other kind of test is an integration or system test, sometimes called a full-stack test, which tests the whole system working as it does in the real world. This kind of test serves a different function -- to confirm that all parts work together in the way the user will expect. This type of test is not as good at figuring out /what's wrong/ -- lower level component (or "unit tests") are best for this. But they can tip us off if /something/ is going wrong, and are better at detecting unexpected failure modes or incorrect assumptions! does this make sense? 

🎉 

As to mocking the related function is not needed - if we set a local destination for this option `module.options.fetchRelated`, it will fetch it from a local relative path, instead of a remote server, and the tests will be able to run:

https://github.com/publiclab/PublicLab.Editor/blob/39bd6e979920a5cb07dc318450fa19dcb68e36d7/src/modules/PublicLab.TitleModule.Related.js#L60-L67

See how, actually, it's already incorporating a mini version of that file as the default fallback for that function. So you could try Jest testing against that mini mocked response that's in the default, even!

**Comment by NitinBhasneria** [comment]
> See how, actually, it's already incorporating a mini version of that file as the default fallback for that function. So you could try Jest testing against that mini mocked response that's in the default, even!


Yea, the example of the mock is provided. But it is not used as when the fetch related function is called it fetches the related title and if it gets `null` it returns the null and replaces this string to null.

Secondly, I am unable to mock the showRelatedResults because it is as `function inside function` and so this cant be tested. The only way to test is( which I came to know so far) is copying the code to test but then it indirectly tested the code not directly. It just tests the copied code, not the real one. Also, we won't be able to test the token field theirs. 
So, what should we do then??😅

**Comment by jywarren** [comment]
Hmm, maybe i misunderstand - couldn't we just not override this function, but instead assert that it is correctly inserting stuff into the page, from the mocked JSON provided? 

https://github.com/publiclab/PublicLab.Editor/blob/39bd6e979920a5cb07dc318450fa19dcb68e36d7/src/modules/PublicLab.TitleModule.Related.js#L41

Is one issue that we have to use a test title term that will actually match some of the mock results?

**Comment by jywarren** [comment]
Maybe @NitinBhasneria could you clarify what you want to test by mocking `show()`?

**Comment by jywarren** [comment]
thank you!!!!

**Comment by NitinBhasneria** [comment]
> Maybe @NitinBhasneria could you clarify what you want to test by mocking `show()`?

We have to give the dummy data of related titles with `show()`.
@jywarren I have found that it is more accurate to add the test in plots2 for the same.

File: spec/javascripts/title_module_spec.js (hunk 16 above)
Comment (jywarren) [reply spec/javascripts/title_module_spec.js:21]: Hi Nitin, are you able to try typing one of the terms in the autocomplete mock JSON that's already in the repo, and then confirming that we can add a related item?

**Comment by NitinBhasneria** [comment]
> Yea, the example of the mock is provided. But it is not used as when the fetch related function is called it fetches the related title and if it gets `null` it returns the null and replaces this string to null.


@jywarren as I have mentioned here. The mock(related titles) get null when it fetches. And so no autocomplete mock is present.

**Comment by sagarpreet-chadha** [comment]
Hi @jywarren , this is pending from long time now. Should we make a separate issue for writing tests here?
@NitinBhasneria I am not sure I understand you completely.

**Comment by NitinBhasneria** [comment]
> Hi @jywarren , this is pending from long time now. Should we make a separate issue for writing tests here?
> @NitinBhasneria I am not sure I understand you completely.

Yea, I think this is a great idea. Separating the tests will give us the range to test it in `editor` plus `plots2` then we can decide which is a more accurate and better way to test. Thank you

**Comment by jywarren** [comment]
OK, i think that's OK - but what you can do is to write a test that fails, and demonstrates the issue you're seeing with the mocked titles returning `null` -- and then we can try to fix it, which will resolve the test. How does that sound? Can you create a new PR demonstrating the issue? Thank you!!!

**Comment by jywarren** [comment]
You know, I am just thinking, it will be really excellent to have the full tests all in one PR. @NitinBhasneria can you rebase this, and add the test that returns `null` and I can put in some time to figure out what's going wrong with it, and we'll merge the whole thing together? Tests are a great way to demonstrate something happening wrong, and we can iterate on it within this PR. Thank you, don't worry, we're very close!!

File: spec/javascripts/title_module_spec.js [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    expect(module.value()).toBe('My new post:');
+
+    expect(module.valid()).toBe(true);
+  });
```
Comment (jywarren) [inline spec/javascripts/title_module_spec.js:25]:
```suggestion
    $('.ple-title-related .result .add-tag').click();
    expect($('.ple-title-related . addedresult').length).toBe(1); // checks that it was indeed added
  });
```

**Comment by NitinBhasneria** [comment]
> How about something like this? But, we can keep going to remove items too, and also add more checks that the item was removed from the overall results at the same time as it was added as an `.addedresult`-- make sense?

I will try this. Thanks

**Comment by NitinBhasneria** [comment]
> You know, I am just thinking, it will be really excellent to have the full tests all in one PR. @NitinBhasneria can you rebase this, and add the test that returns `null` and I can put in some time to figure out what's going wrong with it, and we'll merge the whole thing together? Tests are a great way to demonstrate something happening wrong, and we can iterate on it within this PR. Thank you, don't worry, we're very close!!

yea that's fine.

**Comment by NitinBhasneria** [comment]
@jywarren I have implemented what you have suggested and it worked.

File: spec/javascripts/title_module_spec.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    expect(module.value()).not.toBe(false);
+    expect(module.value()).toBe('');
+
+    module.value('related post');
+    expect(module.value()).toBe('related post'); // should match 3 mocked results
+
+    expect(module.valid()).toBe(true);
+    expect($('.ple-title-related .result').length).toBe(3);  });
+
+    // $('.ple-title-related .result .add-tag').click();
```
Comment (sagarpreet-chadha) [inline spec/javascripts/title_module_spec.js:27]: Let's remove these comments if not needed

File: spec/javascripts/title_module_spec.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    fixture = loadFixtures('index.html');
+
+    editor = new PL.Editor({
+      textarea: $('.ple-textarea')[0]
+    });
+
+    module = new PL.TitleModule(editor, {});
+  });
+
+  it("value, valid, key", function() {
```
Comment (sagarpreet-chadha) [inline spec/javascripts/title_module_spec.js:14]: Can we write properly here what this test is actually doing so that other contributors can easily figure out, thanks 😄

**Review by sagarpreet-chadha — CHANGES_REQUESTED** [review]
small changes and it is good to go 💯
